android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/ssdp/Ssdp.java

```diff
@@ -0,0 +1,140 @@
+ * limitations under the License.
+ */
+
+package org.physical_web.physicalweb.ssdp;
+
+import android.util.Log;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+
+/**
+ * This class implements SSDP protocol.  
+ * It supports discovery of SSDP devices in
+ * the local network using SSDP M-SEARCH search requests
+ */
+public class Ssdp implements Runnable {
+    public static final String TAG = "Ssdp";
+    public static final String SSDP_ADDRESS = "239.255.255.250";
+    public static final int SSDP_PORT = 1900;
+    public static final String SSDP_HOST = SSDP_ADDRESS + ":" + SSDP_PORT;
+    public static final String MAX_AGE = "max-age=1800";
+    public static final int TTL = 128;
+    public static final int MX = 2;
+    public static final String ALIVE = "ssdp:alive";
+    public static final String BYEBYE = "ssdp:byebye";
+    public static final String UPDATE = "ssdp:update";
+    public static final String TYPE_M_SEARCH = "M-SEARCH";
+    public static final String TYPE_NOTIFY = "NOTIFY";
+    public static final String TYPE_200_OK = "200 OK";
+
+    private SsdpCallback mSsdpCallback;
+    private SocketAddress mMulticastGroup;
+    private DatagramSocket mDatagramSocket;
+    private Thread mThread;
+    public Ssdp(SsdpCallback ssdpCallback) throws IOException{
+        mSsdpCallback = ssdpCallback;
+        mMulticastGroup = new InetSocketAddress(SSDP_ADDRESS,SSDP_PORT);
+    }
+
+    public synchronized boolean start(Integer timeout) throws IOException{
+        if (mThread == null){
+            mDatagramSocket = new DatagramSocket(null); // create a DatagramSocket without binding to any address
+            mDatagramSocket.setReuseAddress(true);
+            mDatagramSocket.bind(null); // bind to any free port
+            if (timeout != null && timeout>0){
+                mDatagramSocket.setSoTimeout(timeout);
+            }
+            mThread = new Thread(this);
+            mThread.start();
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized boolean stop() throws IOException{
+        if(mThread != null){
+            mThread.interrupt();
+            mDatagramSocket.close();
+            mThread = null;
+            mDatagramSocket = null;
+            return true;
+        }
+        return false;
+    }
+
+    public void search(SsdpMessage msg) throws IOException{
+        if(mDatagramSocket != null){
+            byte bytes[] = msg.toString().getBytes();
+            DatagramPacket dp = new DatagramPacket(bytes, bytes.length, this.mMulticastGroup);
+            mDatagramSocket.send(dp);
+        }
+    }
+
+    public SsdpMessage search(String ST) throws IOException{
+        SsdpMessage msg = new SsdpMessage(SsdpMessage.TYPE_SEARCH);
+        msg.getHeaders().put("ST",ST);
```

Comment:
Please separate the arguments with a space after each comma:

msg.getHeaders().put("ST", ST);
msg.getHeaders().put("HOST", SSDP_HOST);
msg.getHeaders().put("MAN", "\"ssdp:discover\"");
msg.getHeaders().put("MX", MX+"");